File: sim/simulation.go
package sim

import (
	"fmt"
	"math"
	"math/rand"
	"sync"

	"economy/agent"
	"economy/citizen"
	"economy/company"
	"economy/country"
	"economy/country/fiscal"
	"economy/market"
)

// AccountHolder is satisfied by any participant that exposes its
// underlying market.Account — every concrete participant type built
// so far (agent.MarketMaker, agent.HedgeFund, agent.Bank,
// retail.SimulatedRetailTrader, retail.HumanTrader) already has an
// Account() method, so this costs nothing to require and is what lets
// the orchestrator run a single, global liquidation scan across every
// account in the simulation regardless of participant type. Defined
// locally in sim rather than in market, since market intentionally
// requires nothing beyond OrderSource from participants — this is an
// orchestration-level need, not a market-mechanics one.
type AccountHolder interface {
	market.OrderSource
	Account() *market.Account
}

// Simulation is the orchestrator: it owns every company's market,
// every participant, and drives the tick loop that is the only place
// in this codebase where package market, company, technicals, agent,
// and retail are all used together.
type Simulation struct {
	mu sync.RWMutex

	markets map[string]*companyMarket // symbol -> that company's market
	order   []string                  // symbols in a stable, deterministic iteration order

	holders []AccountHolder // every account-bearing participant, for the global liquidation scan
	banks   []*agent.Bank   // tracked separately so SetExternalPrices can be called each tick — see step 6 in Step's doc

	depthLevels int
	maxHistory  int

	tick int

	// EventLog collects notable events (fundamental jumps,
	// restatements, liquidations) across the run for later
	// inspection — see events.go.
	EventLog []Event
	rng      *rand.Rand

	// Macroeconomic Ecosystem
	Citizen        *citizen.CitizenEconomy
	National       *country.NationalEconomy
	CitizenReport  citizen.CitizenReport
	NationalReport country.NationalReport
}

// NewSimulation constructs an empty orchestrator ready to have
// companies and participants registered via AddCompany/AddParticipant
// before Step is called. depthLevels controls how many order book
// price levels MarketState.BidDepth/AskDepth sum over (see
// market.OrderBook.DepthAtLevels); maxHistory bounds how many past
// mid-prices each company's rolling history retains.
func NewSimulation(depthLevels, maxHistory int) *Simulation {
	cit := citizen.NewDefaultCitizenEconomy()
	nat := country.NewDefaultNationalEconomy()

	// Initial reports
	citRep := cit.Tick(nat.Labor.EmployedWorkers, nat.Labor.AverageHourlyWage, nat.CentralBank.CPIInflationRate, nat.Labor.AnnualWageGrowth, 0.0, 1.0/252.0)
	natRep := nat.Tick(citRep.ConsumerSpending, 5_000_000_000_000.0, 160_000_000.0, 500_000_000_000.0, 4_000_000_000_000.0, 1.0/252.0)

	return &Simulation{
		markets:        make(map[string]*companyMarket),
		depthLevels:    depthLevels,
		maxHistory:     maxHistory,
		rng:            rand.New(rand.NewSource(1337)),
		Citizen:        cit,
		National:       nat,
		CitizenReport:  citRep,
		NationalReport: natRep,
	}
}

// AddCompany registers a company with the simulation, giving it a
// fresh, empty order book. Must be called before any participant that
// trades this company is added, and before the first Step.
func (s *Simulation) AddCompany(co *company.Company) {
	if _, exists := s.markets[co.Symbol]; exists {
		return // idempotent: re-adding an already-registered company is a no-op, not an error
	}
	s.markets[co.Symbol] = newCompanyMarket(co, s.maxHistory)
	s.order = append(s.order, co.Symbol)
}

// AddParticipant registers a participant to trade the given symbols.
// The participant must implement AccountHolder — every concrete
// participant type in this codebase already does. For symbols not yet
// registered via AddCompany, AddParticipant panics rather than
// silently dropping the association, since a participant that
// believes it's trading a company the orchestrator has never heard of
// is a wiring bug worth catching immediately, not a scenario to
// tolerate quietly.
//
// If p is an *agent.Bank, it is additionally tracked so Step can call
// SetExternalPrices on it each tick with real cross-company market
// prices — see Step's documentation, and agent.Bank's own
// documentation on why that matters for accurate drawdown tracking.
func (s *Simulation) AddParticipant(p AccountHolder, symbols ...string) {
	for _, symbol := range symbols {
		cm, ok := s.markets[symbol]
		if !ok {
			panic("sim: AddParticipant called with unregistered symbol " + symbol + " — call AddCompany first")
		}
		cm.participants = append(cm.participants, p)
	}
	s.holders = append(s.holders, p)

	if bank, ok := p.(*agent.Bank); ok {
		s.banks = append(s.banks, bank)
	}
}

// Tick reports the number of completed Step calls so far.
func (s *Simulation) Tick() int {
	return s.tick
}

// Company returns the registered company for symbol, or nil if it
// isn't registered — exposed mainly for tests and result inspection
// (e.g. reading a company's final TrueValue/ReportedValue after a run).
func (s *Simulation) Company(symbol string) *company.Company {
	cm, ok := s.markets[symbol]
	if !ok {
		return nil
	}
	return cm.co
}

// Book returns the registered order book for symbol, or nil if it
// isn't registered — exposed mainly for tests and result inspection.
func (s *Simulation) Book(symbol string) *market.OrderBook {
	cm, ok := s.markets[symbol]
	if !ok {
		return nil
	}
	return cm.book
}

// Symbols returns every registered company's symbol, in the stable
// order they were added.
func (s *Simulation) Symbols() []string {
	out := make([]string, len(s.order))
	copy(out, s.order)
	return out
}

// ListIPO admits a newly generated company into the live simulation as an IPO.
// It registers the company, provisions an underwriter MarketMaker, seeds the opening book
// with institutional depth around the IPO offering price, and records an IPO event.
func (s *Simulation) ListIPO(co *company.Company, underwriterCapital float64) {
	if _, exists := s.markets[co.Symbol]; exists {
		return
	}
	s.AddCompany(co)

	if underwriterCapital <= 0 {
		underwriterCapital = 25_000_000
	}

	// 1. Provision Underwriting Market Maker with dynamic depth
	mm := agent.NewMarketMaker("underwriter_"+co.Symbol, underwriterCapital, 10, 0.08)
	mm.MaxInventory = 500_000
	mm.QuoteSize = 250
	mm.LadderLevels = 5
	mm.SetCompany(co)
	s.AddParticipant(mm, co.Symbol)

	// 2. Provision Institutional Hedge Fund allocating capital
	hfAcct := market.NewAccount("hf_ipo_"+co.Symbol, 15_000_000, 5, 0.10)
	hf := agent.NewHedgeFund("hf_ipo_"+co.Symbol, hfAcct, co)
	hf.MinTradeThreshold = 0.003
	hf.FullConvictionThreshold = 0.025
	hf.RebalanceThreshold = 0.003
	hf.ExecutionRate = 0.06
	hf.MaxOrderShares = 200
	s.AddParticipant(hf, co.Symbol)

	// 3. Provision Institutional Investment Bank covering the company
	bankAcct := market.NewAccount("bank_ipo_"+co.Symbol, 25_000_000, 3, 0.15)
	bank := agent.NewBank("bank_ipo_"+co.Symbol, bankAcct, []*company.Company{co})
	bank.MinTradeThreshold = 0.006
	bank.FullConvictionThreshold = 0.040
	bank.RebalanceThreshold = 0.006
	bank.ExecutionRate = 0.05
	bank.MaxOrderShares = 250
	s.AddParticipant(bank, co.Symbol)

	offeringPrice := co.TrueValue
	if offeringPrice <= 0 {
		offeringPrice = 100.0
	}

	// Seed opening quotes from the market maker so newly listed IPO has immediate valid mid price
	book := s.Book(co.Symbol)
	for i := range 5 {
		offset := 0.05 * float64(i+1)
		book.AddLimitOrder(&market.Order{AgentID: mm.ID(), Side: market.Buy, Price: math.Round((offeringPrice-offset)*100) / 100, Quantity: 250})
		book.AddLimitOrder(&market.Order{AgentID: mm.ID(), Side: market.Sell, Price: math.Round((offeringPrice+offset)*100) / 100, Quantity: 250})
	}

	// 4. Record IPO event
	s.EventLog = append(s.EventLog, Event{
		Tick:       s.tick,
		Kind:       EventIPO,
		Symbol:     co.Symbol,
		IPORevenue: co.AnnualRevenue,
		IPOPrice:   offeringPrice,
		IPOShares:  co.SharesOutstanding,
	})
}

// SetTick sets the current tick (used when restoring saved state).
func (s *Simulation) SetTick(t int) {
	s.tick = t
}

// BailoutCompany executes a government rescue facility for a distressed corporation:
// injects capital, slashes debt by 50%, cuts interest burden, and lifts valuation by +50%.
func (s *Simulation) BailoutCompany(symbol string, capitalInjection float64) error {
	cm, ok := s.markets[symbol]
	if !ok {
		return fmt.Errorf("company %s not registered in simulation", symbol)
	}
	co := cm.co
	if capitalInjection <= 0 {
		capitalInjection = 500_000_000.0
	}

	debtRelief := co.DebtOutstanding * 0.50
	co.DebtOutstanding -= debtRelief
	co.InterestExpense *= 0.50
	co.TrueValue *= 1.50
	co.ReportedValue *= 1.50

	if s.National != nil {
		s.National.Fiscal.NationalDebt += capitalInjection
	}

	headline := fmt.Sprintf("Federal Government injected $%.0fM emergency facility (cut debt by $%.0fM, +50%% intrinsic value)", capitalInjection/1e6, debtRelief/1e6)
	s.EventLog = append(s.EventLog, Event{
		Tick:            s.tick,
		Kind:            EventBailout,
		Symbol:          symbol,
		DistressDetails: headline,
	})

	return nil
}

// TogglePolicy flips an executive policy on or off, emits an announcement event,
// and applies direct stimulus effects if the citizen stimulus policy is enacted.
func (s *Simulation) TogglePolicy(policyID string) (bool, string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.National == nil {
		return false, "National economy not initialized"
	}
	active, name := s.National.Fiscal.TogglePolicy(policyID)

	// If citizen stimulus was triggered, inject directly into Citizen demographics and sentiment!
	if policyID == fiscal.PolicyCitizenStimulus && active && s.Citizen != nil {
		s.Citizen.Sentiment.Index = math.Min(100, s.Citizen.Sentiment.Index+15.0)
		s.Citizen.Sentiment.Happiness = math.Min(100, s.Citizen.Sentiment.Happiness+12.0)
		s.Citizen.Demographics.AggregateDisposableIncome += 320_000_000_000.0
		s.National.Fiscal.NationalDebt += 320_000_000_000.0
	}

	actionStr := "ENACTED"
	if !active {
		actionStr = "REPEALED"
	}
	headline := fmt.Sprintf("Executive Policy %s: %s", actionStr, name)
	s.EventLog = append(s.EventLog, Event{
		Tick:          s.tick,
		Kind:          EventMacro,
		Symbol:        "US_GOV",
		MacroHeadline: headline,
	})

	if s.National != nil {
		s.NationalReport.ActivePolicies = s.National.Fiscal.GetActivePolicies()
	}

	return active, name
}

File: src/sim/simulation.rs
use std::collections::HashMap;
use std::sync::Arc;
use parking_lot::RwLock;
use rand::rngs::StdRng;
use rand::SeedableRng;

use crate::agent::{Bank, HedgeFund, MarketMaker};
use crate::citizen::{CitizenEconomy, CitizenReport};
use crate::company::Company;
use crate::country::{all_known_policies, NationalEconomy, NationalReport, POLICY_CITIZEN_STIMULUS};
use crate::market::{Account, Order, OrderBook, OrderSource, Side};
use super::company_market::CompanyMarket;
use super::events::{Event, EventKind};

pub trait AccountHolder: OrderSource {
    fn account(&self) -> Arc<RwLock<Account>>;
}

impl AccountHolder for MarketMaker {
    fn account(&self) -> Arc<RwLock<Account>> {
        self.account.clone()
    }
}

impl AccountHolder for HedgeFund {
    fn account(&self) -> Arc<RwLock<Account>> {
        self.account.clone()
    }
}

impl AccountHolder for Bank {
    fn account(&self) -> Arc<RwLock<Account>> {
        self.account.clone()
    }
}

impl AccountHolder for crate::retail::SimulatedRetailTrader {
    fn account(&self) -> Arc<RwLock<Account>> {
        self.account.clone()
    }
}

impl AccountHolder for crate::retail::HumanTrader {
    fn account(&self) -> Arc<RwLock<Account>> {
        self.account.clone()
    }
}

pub struct Simulation {
    pub markets: HashMap<String, CompanyMarket>,
    pub order: Vec<String>,
    pub holders: Vec<Arc<RwLock<dyn AccountHolder>>>,
    pub banks: Vec<Arc<RwLock<Bank>>>,

    pub depth_levels: usize,
    pub max_history: usize,
    pub tick: usize,

    pub event_log: Vec<Event>,
    pub rng: StdRng,

    pub citizen: CitizenEconomy,
    pub national: NationalEconomy,
    pub citizen_report: CitizenReport,
    pub national_report: NationalReport,
    pub private_companies: Vec<Company>,
}

impl Simulation {
    pub fn new(depth_levels: usize, max_history: usize) -> Self {
        let mut cit = CitizenEconomy::new();
        let mut nat = NationalEconomy::new();

        let cit_rep = cit.tick(
            nat.labor.employed_workers,
            nat.labor.average_hourly_wage,
            nat.central_bank.cpi_inflation_rate,
            nat.labor.annual_wage_growth,
            0.0,
            1.0 / 252.0,
        );
        let nat_rep = nat.tick(
            cit_rep.consumer_spending,
            3_000_000_000.0,
            4_000_000.0,
            250_000_000.0,
            350_000_000.0,
            1.0 / 252.0,
        );

        Self {
            markets: HashMap::new(),
            order: Vec::new(),
            holders: Vec::new(),
            banks: Vec::new(),
            depth_levels,
            max_history,
            tick: 0,
            event_log: Vec::new(),
            rng: StdRng::seed_from_u64(1337),
            citizen: cit,
            national: nat,
            citizen_report: cit_rep,
            national_report: nat_rep,
            private_companies: Vec::new(),
        }
    }

    pub fn add_company(&mut self, co: Company) {
        if self.markets.contains_key(&co.symbol) {
            return;
        }
        let symbol = co.symbol.clone();
        self.markets.insert(symbol.clone(), CompanyMarket::new(co, self.max_history));
        self.order.push(symbol);
    }

    pub fn add_private_company(&mut self, co: Company) {
        self.private_companies.push(co);
    }

    pub fn add_participant<P: AccountHolder + 'static>(
        &mut self,
        p: Arc<RwLock<P>>,
        symbols: &[&str],
    ) {
        for &sym in symbols {
            let cm = self.markets.get_mut(sym).expect("sim: AddParticipant called with unregistered symbol");
            cm.participants.push(p.clone());
        }
        self.holders.push(p);
    }

    pub fn add_bank(&mut self, bank: Arc<RwLock<Bank>>, symbols: &[&str]) {
        for &sym in symbols {
            let cm = self.markets.get_mut(sym).expect("sim: AddParticipant called with unregistered symbol");
            cm.participants.push(bank.clone());
        }
        self.holders.push(bank.clone());
        self.banks.push(bank);
    }

    pub fn tick(&self) -> usize {
        self.tick
    }

    pub fn set_tick(&mut self, t: usize) {
        self.tick = t;
    }

    pub fn company(&self, symbol: &str) -> Option<&Company> {
        self.markets.get(symbol).map(|cm| &cm.co)
    }

    pub fn company_mut(&mut self, symbol: &str) -> Option<&mut Company> {
        self.markets.get_mut(symbol).map(|cm| &mut cm.co)
    }

    pub fn book(&self, symbol: &str) -> Option<&OrderBook> {
        self.markets.get(symbol).map(|cm| &cm.book)
    }

    pub fn book_mut(&mut self, symbol: &str) -> Option<&mut OrderBook> {
        self.markets.get_mut(symbol).map(|cm| &mut cm.book)
    }

    pub fn symbols(&self) -> Vec<String> {
        self.order.clone()
    }

    pub fn clear_event_log(&mut self) {
        self.event_log.clear();
    }

    pub fn list_ipo(&mut self, co: Company, underwriter_capital: f64) {
        if self.markets.contains_key(&co.symbol) {
            return;
        }

        let symbol = co.symbol.clone();
        let capital = if underwriter_capital <= 0.0 { 25_000_000.0 } else { underwriter_capital };

        self.add_company(co.clone());

        let mut mm = MarketMaker::new(format!("underwriter_{}", symbol), capital, 10.0, 0.08);
        mm.max_inventory = 500_000.0;
        mm.quote_size = 250.0;
        mm.ladder_levels = 5;
        mm.set_company(co.clone());
        let mm_arc = Arc::new(RwLock::new(mm));
        self.add_participant(mm_arc.clone(), &[&symbol]);

        let hf_acct = Arc::new(RwLock::new(Account::new(
            format!("hf_ipo_{}", symbol),
            15_000_000.0,
            5.0,
            0.10,
        )));
        let mut hf = HedgeFund::new(format!("hf_ipo_{}", symbol), hf_acct, co.clone());
        hf.min_trade_threshold = 0.003;
        hf.full_conviction_threshold = 0.025;
        hf.rebalance_threshold = 0.003;
        hf.execution_rate = 0.06;
        hf.max_order_shares = 200.0;
        let hf_arc = Arc::new(RwLock::new(hf));
        self.add_participant(hf_arc, &[&symbol]);

        let bank_acct = Arc::new(RwLock::new(Account::new(
            format!("bank_ipo_{}", symbol),
            25_000_000.0,
            3.0,
            0.15,
        )));
        let mut bank = Bank::new(format!("bank_ipo_{}", symbol), bank_acct, vec![co.clone()]);
        bank.min_trade_threshold = 0.006;
        bank.full_conviction_threshold = 0.040;
        bank.rebalance_threshold = 0.006;
        bank.execution_rate = 0.05;
        bank.max_order_shares = 250.0;
        let bank_arc = Arc::new(RwLock::new(bank));
        self.add_bank(bank_arc, &[&symbol]);

        let offering_price = if co.true_value <= 0.0 { 100.0 } else { co.true_value };
        let book = self.book_mut(&symbol).unwrap();
        for i in 0..5 {
            let offset = 0.05 * (i as f64 + 1.0);
            let bp = ((offering_price - offset) * 100.0).round() / 100.0;
            let ap = ((offering_price + offset) * 100.0).round() / 100.0;
            book.add_limit_order(Order {
                id: 0,
                agent_id: format!("underwriter_{}", symbol),
                side: Side::Buy,
                price: bp,
                quantity: 250.0,
                is_market: false,
            });
            book.add_limit_order(Order {
                id: 0,
                agent_id: format!("underwriter_{}", symbol),
                side: Side::Sell,
                price: ap,
                quantity: 250.0,
                is_market: false,
            });
        }

        self.event_log.push(Event {
            tick: self.tick,
            kind: EventKind::IPO,
            symbol: symbol.clone(),
            fundamental_kind: None,
            multiplier: None,
            restatement_profile: None,
            prior_gap_percent: None,
            severity: None,
            liquidated_agent_id: None,
            liquidation_qty: None,
            ipo_revenue: Some(co.annual_revenue),
            ipo_price: Some(offering_price),
            ipo_shares: Some(co.shares_outstanding),
            distress_details: None,
            macro_headline: None,
        });
    }

    pub fn bailout_company(&mut self, symbol: &str, mut capital_injection: f64) -> Result<(), &'static str> {
        let cm = self.markets.get_mut(symbol).ok_or("company not registered in simulation")?;
        if capital_injection <= 0.0 {
            capital_injection = 500_000_000.0;
        }

        let debt_relief = cm.co.debt_outstanding * 0.50;
        cm.co.debt_outstanding -= debt_relief;
        cm.co.interest_expense *= 0.50;
        cm.co.true_value *= 1.50;
        cm.co.reported_value *= 1.50;

        self.national.fiscal.national_debt += capital_injection;

        let headline = format!(
            "Federal Government injected ${:.0}M emergency facility (cut debt by ${:.0}M, +50% intrinsic value)",
            capital_injection / 1e6,
            debt_relief / 1e6
        );

        self.event_log.push(Event {
            tick: self.tick,
            kind: EventKind::Bailout,
            symbol: symbol.to_string(),
            fundamental_kind: None,
            multiplier: None,
            restatement_profile: None,
            prior_gap_percent: None,
            severity: None,
            liquidated_agent_id: None,
            liquidation_qty: None,
            ipo_revenue: None,
            ipo_price: None,
            ipo_shares: None,
            distress_details: Some(headline),
            macro_headline: None,
        });

        Ok(())
    }

    pub fn toggle_policy(&mut self, policy_id: &str) -> Result<(bool, String), String> {
        let (tier, _, _) = self.national.current_tier();
        let (active, name) = self.national.fiscal.toggle_policy_checked(policy_id, tier)?;

        if policy_id == POLICY_CITIZEN_STIMULUS && active {
            self.citizen.sentiment.index = (self.citizen.sentiment.index + 15.0).min(100.0);
            self.citizen.sentiment.happiness = (self.citizen.sentiment.happiness + 12.0).min(100.0);
            self.citizen.demographics.aggregate_disposable_income += 12_000_000_000.0;
            self.national.fiscal.national_debt += 12_000_000_000.0;
        }

        // Immediate market shock: apply sector valuation impact on toggle
        let policies = all_known_policies();
        if let Some(pol_info) = policies.iter().find(|p| p.id == policy_id) {
            // The shock is 0.5% of the procurement delta per company in the favored sector
            // Policy active = positive shock, deactivated = negative shock
            let shock_mult = if active { 1.005_f64 } else { 0.995_f64 };
            let favored = &pol_info.favored_sector;
            for symbol in &self.order {
                if let Some(cm) = self.markets.get_mut(symbol) {
                    let sector_name = cm.co.sector.to_string();
                    if favored == "All Sectors" || &sector_name == favored {
                        cm.co.true_value = (cm.co.true_value * shock_mult).max(0.01);
                        cm.co.reported_value = (cm.co.reported_value * shock_mult).max(0.01);
                    }
                }
            }
        }

        let action_str = if active { "ENACTED" } else { "REPEALED" };
        let headline = format!("Executive Policy {}: {}", action_str, name);

        self.event_log.push(Event {
            tick: self.tick,
            kind: EventKind::Macro,
            symbol: "SOVEREIGN_GOV".to_string(),
            fundamental_kind: None,
            multiplier: None,
            restatement_profile: None,
            prior_gap_percent: None,
            severity: None,
            liquidated_agent_id: None,
            liquidation_qty: None,
            ipo_revenue: None,
            ipo_price: None,
            ipo_shares: None,
            distress_details: None,
            macro_headline: Some(headline),
        });

        self.national_report.active_policies = self.national.fiscal.get_active_policies();
        Ok((active, name))
    }
}
